src/Services/Qrda/QrdaParseService.php

```diff
@@ -0,0 +1,315 @@
+            //'2.16.840.1.113883.10.20.24.3.139' => 'fetchMedicationData', //Medication Dispensed Act
+            //'2.16.840.1.113883.10.20.24.3.42' => 'fetchMedicationData', // prescribe QDM Datatype: Substance, Administered
+            //'2.16.840.1.113883.10.20.24.3.47' => 'fetchMedicationData', // planned QDM Datatype: Substance, Order i.e prescribed
+            //'2.16.840.1.113883.10.20.22.4.3' => '', //cda concern act
+            '2.16.840.1.113883.10.20.24.3.137' => 'fetchMedicalProblemData', //qrda diagnosis
+            '2.16.840.1.113883.10.20.24.3.140' => 'fetchImmunizationData', // qrda Immunization Administered (V3)
+            '2.16.840.1.113883.10.20.24.3.143' => '', // qrda Immunization Order (V3)
+            //'2.16.840.1.113883.10.20.22.4.14' => '', // C-CDA R2.1 Procedure Activity Procedure (V2)
+            '2.16.840.1.113883.10.20.24.3.64' => 'fetchProcedureData', // qrda procedure performed
+            '2.16.840.1.113883.10.20.24.3.7' => 'fetchProcedureData', // qrda procedure Device Applied (V5)
+            //'2.16.840.1.113883.10.20.24.3.38' => 'lab_result', // lab test preformed wip
+            //'2.16.840.1.113883.10.20.24.3.37' => 'lab_result', // lab test ordered wip
+            '2.16.840.1.113883.10.20.24.3.133' => 'fetchEncounter',
+        );
+        foreach ($entryComponents['section']['entry'] as $i => $entry) {
+            $key = array_keys($entry)[0]; // need the entry type i.e. observation, activity, substance etc.
+            if (!empty($entry[$key]['templateId']['root'])) {
+                if (!empty($qrda_oids[$entry[$key]['templateId']['root']])) {
+                    $func_name = $qrda_oids[$entry[$key]['templateId']['root']];
+                    $this->$func_name($entry);
+                }
+            } elseif (count($entry[$key]['templateId']) > 1) {
+                foreach ($entry[$key]['templateId'] as $key_1 => $value_1) {
+                    if (!empty($qrda_oids[$entry[$key]['templateId'][$key_1]['root']])) {
+                        $func_name = $qrda_oids[$entry[$key]['templateId'][$key_1]['root']];
+                        if (!empty($func_name)) {
+                            $this->$func_name($entry);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+        return $this->qrdaData;
+    }
+
+    /**
+     * @param $entry
+     */
+    public function fetchEncounter($entry): void
+    {
+        $this->getEncounterData($entry['act']['entryRelationship']);
+    }
+
+    /**
+     * @param $entry
+     */
+    public function getEncounterData($entry): void
+    {
+        if ($entry['encounter']['effectiveTime']['value'] != 0 || $entry['encounter']['effectiveTime']['low']['value'] != 0) {
+            $i = 1;
+            if (!empty($this->qrdaData['field_name_value_array']['encounter'])) {
+                $i += count($this->qrdaData['field_name_value_array']['encounter']);
```

Comment:
This is copied from ccda right?  This essentially makes it so we add a new encounter record to the $this->qrdaData['field_name_value_array']['encounter'] right?

So strange of ZH to do it this way.

Reply:
The qrdaData is a holding array of all template parsed data with the reason they chose this is so we can audit the import for user to review and approve. At that point the controller inserts/appends or updates the chart from the array. See screen shot above.
Yes most methods came from the ccda however their x-pathing is different and qrda template will require more parsing than existing ccda import templates. Besides, CCM import is messy w/o adding to the mess:)

My plan is to further dev the new service methods to return their results where parsePatientDataSection will collect and format for CCM import but still allow calling individual methods for eventual auto parsing engine...